domains/client/domain-executor/src/core_bundle_processor.rs

```diff
@@ -0,0 +1,268 @@
+            keystore,
+            core_domain_block_preprocessor,
+            domain_block_processor,
+        }
+    }
+
+    // TODO: Handle the returned error properly, ref to https://github.com/subspace/subspace/pull/695#discussion_r926721185
+    pub(crate) async fn process_bundles(
+        self,
+        primary_info: (PBlock::Hash, NumberFor<PBlock>, bool),
+    ) -> sp_blockchain::Result<()> {
+        let (primary_hash, primary_number, is_new_best) = primary_info;
+
+        tracing::debug!("Processing imported primary block #{primary_number},{primary_hash}");
+
+        let maybe_pending_primary_blocks = self
+            .domain_block_processor
+            .pending_imported_primary_blocks(primary_hash, primary_number)?;
+
+        if let Some(PendingPrimaryBlocks {
+            initial_parent,
+            primary_imports,
+        }) = maybe_pending_primary_blocks
+        {
+            tracing::trace!(
+                ?initial_parent,
+                ?primary_imports,
+                "Pending primary blocks to process"
+            );
+
+            let mut domain_parent = initial_parent;
+
+            for primary_info in primary_imports {
+                if let Some(next_domain_parent) = self
+                    .process_bundles_at((primary_info.hash, primary_info.number), domain_parent)
+                    .await?
+                {
+                    domain_parent = next_domain_parent;
+                }
+            }
+
+            // The domain branch driving from the best primary branch should also be the best domain branch even
+            // if it is no the longest domain branch. Thus re-import the tip of the best domain branch to make it
+            // the new best block if it isn't.
+            //
+            // Note: this may cause the best domain fork switch to a shorter fork or in some case the best domain
+            // block will become the ancestor block of the current best block.
+            let domain_tip = domain_parent.0;
+            if is_new_best && self.client.info().best_hash != domain_tip {
+                let header = self.client.header(domain_tip)?.ok_or_else(|| {
+                    sp_blockchain::Error::Backend(format!(
+                        "Header for #{:?} not found",
+                        domain_parent
+                    ))
+                })?;
+                let block_import_params = {
+                    let mut import_block = BlockImportParams::new(BlockOrigin::Own, header);
+                    import_block.import_existing = true;
+                    import_block.fork_choice = Some(ForkChoiceStrategy::Custom(true));
+                    import_block.state_action = StateAction::Skip;
+                    import_block
+                };
+                self.domain_block_processor
+                    .import_domain_block(block_import_params)
+                    .await?;
+                assert_eq!(domain_tip, self.client.info().best_hash);
```

Comment:
do we need to assert here ?

Reply:
It is a kind of defense, to ensure re-import block with `ForkChoiceStrategy::Custom(true)` does make it the new best block.

Reply:
Why? Feels more like we are testing the underlying substrate import logic here.